pkg/atomix/log/service.go

```diff
@@ -0,0 +1,544 @@
+	m.firstEntry = nil
+	m.lastEntry = nil
+	m.indexes = make(map[uint64]*LinkedLogEntryValue)
+	for i := 0; i < entryCount; i++ {
+		value, err := util.ReadValue(reader, func(data []byte) (*LogEntryValue, error) {
+			entry := &LogEntryValue{}
+			if err := proto.Unmarshal(data, entry); err != nil {
+				return nil, err
+			}
+			return entry, nil
+		})
+		if err != nil {
+			return err
+		}
+
+		linkedEntry := &LinkedLogEntryValue{
+			LogEntryValue: value.(*LogEntryValue),
+		}
+		if m.firstEntry == nil {
+			m.firstEntry = linkedEntry
+		}
+
+		m.indexes[linkedEntry.Index] = linkedEntry
+		if prevEntry != nil {
+			prevEntry.Next = linkedEntry
+			linkedEntry.Prev = prevEntry
+		}
+		prevEntry = linkedEntry
+		m.lastEntry = linkedEntry
+	}
+	return nil
+}
+
+// Append appends a key/value pair in the log
+func (m *Service) Append(value []byte) ([]byte, error) {
+	request := &AppendRequest{}
+	if err := proto.Unmarshal(value, request); err != nil {
+		return nil, err
+	}
+
+	var oldEntry *LinkedLogEntryValue
+	if request.Index > 0 {
+		oldEntry = m.indexes[request.Index]
+		if oldEntry != nil && oldEntry.Index != request.Index {
+			return proto.Marshal(&AppendResponse{
+				Status: UpdateStatus_PRECONDITION_FAILED,
+			})
+		}
+	} else {
+		oldEntry = m.indexes[request.Index]
+	}
+
+	if oldEntry == nil {
+		// If the version is positive then reject the request.
+		if !request.IfEmpty && request.Version > 0 {
+			return proto.Marshal(&AppendResponse{
+				Status: UpdateStatus_PRECONDITION_FAILED,
+			})
+		}
+
+		// Increment the index for a new entry
+		var index uint64
+		if request.Index > 0 {
+			if request.Index > m.lastIndex {
+				m.lastIndex = request.Index
+			}
+			index = request.Index
+		} else {
+			m.lastIndex++
+			index = m.lastIndex
+		}
+
+		// Create a new entry value and set it in the log.
+		newEntry := &LinkedLogEntryValue{
+			LogEntryValue: &LogEntryValue{
+				Index:     index,
+				Value:     request.Value,
+				Version:   m.Context.Index(),
+				Timestamp: m.Context.Timestamp(),
+			},
+		}
+		m.indexes[newEntry.Index] = newEntry
+
+		// Set the first entry if not set
+		if m.firstEntry == nil {
+			m.firstEntry = newEntry
+		}
+
+		// If the last entry is set, link it to the new entry
+		if request.Index > 0 {
+			if m.lastIndex == request.Index {
+				if m.lastEntry != nil {
+					m.lastEntry.Next = newEntry
+					newEntry.Prev = m.lastEntry
+				}
+				m.lastEntry = newEntry
+			}
+		} else {
+			if m.lastEntry != nil {
+				m.lastEntry.Next = newEntry
+				newEntry.Prev = m.lastEntry
+			}
+		}
+
+		// Update the last entry
+		m.lastEntry = newEntry
+
+		return proto.Marshal(&AppendResponse{
+			Status: UpdateStatus_OK,
+			Index:  newEntry.Index,
+		})
+	}
+
+	// If the version is -1 then reject the request.
+	// If the version is positive then compare the version to the current version.
+	if request.IfEmpty || (!request.IfEmpty && request.Version > 0 && request.Version != oldEntry.Version) {
```

Comment:
This optimistic lock with `Version` doesn’t make sense within the context of a log. We should remove it